File: python/assistant-stream/src/assistant_stream/serialization/assistant_transport.py
from assistant_stream.assistant_stream_chunk import (
    AnnotationsChunk,
    AssistantStreamChunk,
    FileChunk,
    ReasoningDeltaChunk,
    ReasoningPartStartChunk,
    SourceChunk,
    StepFinishChunk,
    StepStartChunk,
    TextDeltaChunk,
    ToolCallBeginChunk,
    ToolCallDeltaChunk,
    ToolResultChunk,
)
from assistant_stream.serialization.assistant_stream_response import (
    AssistantStreamResponse,
)
from assistant_stream.serialization.heartbeat import (
    SSE_HEARTBEAT_LINE,
    HeartbeatOption,
)
from assistant_stream.serialization.stream_encoder import StreamEncoder
from assistant_stream.serialization.tool_args_settle import ToolCallArgsSettler
from assistant_stream.state_proxy import StateProxyJSONEncoder
from typing import AsyncGenerator, Any
import json
import logging

logger = logging.getLogger(__name__)


class _Canonicalizer:
    """Translate the internal flat chunk dialect into the canonical
    assistant-transport wire shape consumed by the TS AssistantTransportDecoder
    and AssistantMessageAccumulator.

    Each content part is framed by `part-start`/`part-finish` at an allocated
    `path`, streamed text flows as `text-delta` into the open part, and chunk
    type names match the TS `AssistantStreamChunk` union. `update-state` and
    `error` are already message-level and pass through unchanged.
    """

    def __init__(self) -> None:
        self._part_counter = 0
        self._append_part: tuple[str, list[int], str | None] | None = None
        self._tool_paths: dict[str, list[int]] = {}
        self._open_tool_parts: set[str] = set()
        self._tool_args = ToolCallArgsSettler(
            self._finish_tool_call_args,
            self._warn,
            emit_empty_args_text=False,
        )
        self._warned_reasons: set[str] = set()

    def _warn(self, reason: str, detail: str) -> None:
        logger.warning("Dropped assistant-transport chunk (%s): %s", reason, detail)

    def _warn_once(self, reason: str, detail: str) -> None:
        if reason in self._warned_reasons:
            return
        self._warned_reasons.add(reason)
        self._warn(reason, detail)

    def _next_path(self) -> list[int]:
        path = [self._part_counter]
        self._part_counter += 1
        return path

    def _close_append_part(self) -> list[dict[str, Any]]:
        if self._append_part is None:
            return []
        path = self._append_part[1]
        self._append_part = None
        return [{"type": "part-finish", "path": path}]

    def _start_reasoning_part(
        self, chunk: ReasoningPartStartChunk
    ) -> list[dict[str, Any]]:
        # Registered as the append target so the deltas that follow extend this
        # part instead of opening a second one.
        if chunk.unstable_summary is None:
            return []
        frames = self._close_append_part()
        path = self._next_path()
        part: dict[str, Any] = {
            "type": "reasoning",
            "unstable_summary": chunk.unstable_summary,
        }
        if chunk.parent_id is not None:
            part["parentId"] = chunk.parent_id
        frames.append({"type": "part-start", "part": part, "path": []})
        self._append_part = ("reasoning", path, chunk.parent_id)
        return frames

    def _append_delta(
        self, chunk: TextDeltaChunk | ReasoningDeltaChunk, text: str, kind: str
    ) -> list[dict[str, Any]]:
        parent_id = chunk.parent_id
        if (
            self._append_part is None
            or self._append_part[0] != kind
            or self._append_part[2] != parent_id
        ):
            frames = self._close_append_part()
            path = self._next_path()
            part: dict[str, Any] = {"type": kind}
            if parent_id is not None:
                part["parentId"] = parent_id
            frames.append({"type": "part-start", "part": part, "path": []})
            self._append_part = (kind, path, parent_id)
        else:
            frames = []
            path = self._append_part[1]
        frames.append({"type": "text-delta", "textDelta": text, "path": path})
        return frames

    def _begin_tool_call(self, chunk: ToolCallBeginChunk) -> list[dict[str, Any]]:
        if chunk.tool_call_id in self._tool_paths:
            self._warn_once(
                "duplicate-tool-call-id",
                f"tool-call-begin for {chunk.tool_call_id}",
            )
            return []
        frames = self._close_append_part()
        path = self._next_path()
        self._tool_paths[chunk.tool_call_id] = path
        self._open_tool_parts.add(chunk.tool_call_id)
        self._tool_args.begin(chunk.tool_call_id)
        part: dict[str, Any] = {
            "type": "tool-call",
            "toolCallId": chunk.tool_call_id,
            "toolName": chunk.tool_name,
        }
        if chunk.parent_id is not None:
            part["parentId"] = chunk.parent_id
        frames.append({"type": "part-start", "part": part, "path": []})
        return frames

    def _tool_call_delta(self, chunk: ToolCallDeltaChunk) -> list[dict[str, Any]]:
        path = self._tool_paths.get(chunk.tool_call_id)
        if path is None:
            self._warn_once(
                "unknown-tool-call-id",
                f"tool-call-delta for {chunk.tool_call_id}",
            )
            return []
        if not self._tool_args.append(chunk.tool_call_id):
            return []
        return [
            {"type": "text-delta", "textDelta": chunk.args_text_delta, "path": path}
        ]

    def _tool_result(self, chunk: ToolResultChunk) -> list[dict[str, Any]]:
        path = self._tool_paths.get(chunk.tool_call_id)
        result: dict[str, Any] = {
            "type": "result",
            "result": chunk.result,
            "isError": chunk.is_error,
        }
        if chunk.artifact is not None:
            result["artifact"] = chunk.artifact
        if chunk.is_preliminary:
            result["isPreliminary"] = True
        if path is None:
            # A result without a matching tool-call part has no part to address;
            # emit it at the message root so it stays on the wire.
            result["path"] = []
            return [result]
        result["path"] = path
        frames = [result, *self._tool_args.finish(chunk.tool_call_id)]
        if not chunk.is_preliminary:
            frames.extend(self._finish_tool_part(chunk.tool_call_id))
        return frames

    def _finish_tool_call_args(
        self, tool_call_id: str, args_text_delta: str, has_args_text: bool
    ) -> list[dict[str, Any]]:
        path = self._tool_paths.get(tool_call_id)
        if path is None:
            return []
        frames: list[dict[str, Any]] = []
        if args_text_delta:
            frames.append(
                {
                    "type": "text-delta",
                    "textDelta": args_text_delta,
                    "path": path,
                }
            )
        elif not has_args_text:
            frames.append({"type": "text-delta", "textDelta": "{}", "path": path})
        frames.append({"type": "tool-call-args-text-finish", "path": path})
        return frames

    def _finish_tool_part(self, tool_call_id: str) -> list[dict[str, Any]]:
        if tool_call_id not in self._open_tool_parts:
            return []
        self._open_tool_parts.discard(tool_call_id)
        return [{"type": "part-finish", "path": self._tool_paths[tool_call_id]}]

    def _finish_tool_call(
        self, tool_call_id: str, args_text_delta: str = ""
    ) -> list[dict[str, Any]]:
        return [
            *self._tool_args.finish(tool_call_id, args_text_delta),
            *self._finish_tool_part(tool_call_id),
        ]

    def _source(self, chunk: SourceChunk) -> list[dict[str, Any]]:
        frames = self._close_append_part()
        path = self._next_path()
        part: dict[str, Any] = {
            "type": "source",
            "sourceType": chunk.source_type,
            "id": chunk.id,
            "url": chunk.url,
        }
        if chunk.title is not None:
            part["title"] = chunk.title
        if chunk.parent_id is not None:
            part["parentId"] = chunk.parent_id
        frames.append({"type": "part-start", "part": part, "path": []})
        frames.append({"type": "part-finish", "path": path})
        return frames

    def _file(self, chunk: FileChunk) -> list[dict[str, Any]]:
        frames = self._close_append_part()
        path = self._next_path()
        part: dict[str, Any] = {
            "type": "file",
            "data": chunk.data,
            "mimeType": chunk.mime_type,
        }
        if chunk.parent_id is not None:
            part["parentId"] = chunk.parent_id
        frames.append({"type": "part-start", "part": part, "path": []})
        frames.append({"type": "part-finish", "path": path})
        return frames

    def translate(self, chunk: AssistantStreamChunk) -> list[dict[str, Any]]:
        match chunk.type:
            case "text-delta":
                return self._append_delta(chunk, chunk.text_delta, "text")
            case "reasoning-part-start":
                return self._start_reasoning_part(chunk)
            case "reasoning-delta":
                return self._append_delta(chunk, chunk.reasoning_delta, "reasoning")
            case "tool-call-begin":
                return self._begin_tool_call(chunk)
            case "tool-call-delta":
                return self._tool_call_delta(chunk)
            case "tool-call-args-text-finish":
                return self._finish_tool_call(
                    chunk.tool_call_id, chunk.args_text_delta
                )
            case "tool-result":
                return self._tool_result(chunk)
            case "source":
                return self._source(chunk)
            case "file":
                return self._file(chunk)
            case "data":
                # The TS union requires `data` to be an array; wrap single values
                # so a dict payload satisfies the accumulator's array spread.
                return [{"type": "data", "data": [chunk.data]}]
            case "annotations":
                return [{"type": "annotations", "annotations": chunk.annotations}]
            case "step-start":
                return [{"type": "step-start", "messageId": chunk.message_id}]
            case "step-finish":
                return [
                    {
                        "type": "step-finish",
                        "finishReason": chunk.finish_reason,
                        "usage": {
                            "inputTokens": chunk.input_tokens,
                            "outputTokens": chunk.output_tokens,
                        },
                        "isContinued": chunk.is_continued,
                    }
                ]
            case "error":
                return [{"type": "error", "error": chunk.error}]
            case "update-state":
                return [{"type": "update-state", "operations": chunk.operations}]
            case _:
                self._warn_once("unknown-chunk-type", chunk.type)
                return []

    def close(self) -> list[dict[str, Any]]:
        frames = self._close_append_part()
        for tool_call_id in tuple(self._tool_paths):
            frames.extend(self._finish_tool_call(tool_call_id))
        self._tool_paths.clear()
        self._open_tool_parts.clear()
        self._tool_args.clear()
        return frames


class AssistantTransportEncoder(StreamEncoder):
    """
    AssistantTransportEncoder encodes AssistantStreamChunks into the canonical
    assistant-transport SSE wire shape and emits [DONE] when the stream
    completes.
    """

    def get_media_type(self) -> str:
        return "text/event-stream"

    def get_keepalive_token(self) -> str:
        return SSE_HEARTBEAT_LINE

    async def encode_stream(
        self, stream: AsyncGenerator[AssistantStreamChunk, None]
    ) -> AsyncGenerator[str, None]:
        canonicalizer = _Canonicalizer()
        async for chunk in stream:
            for frame in canonicalizer.translate(chunk):
                yield f"data: {json.dumps(frame, cls=StateProxyJSONEncoder)}\n\n"
        for frame in canonicalizer.close():
            yield f"data: {json.dumps(frame, cls=StateProxyJSONEncoder)}\n\n"
        yield "data: [DONE]\n\n"


class AssistantTransportResponse(AssistantStreamResponse):
    def __init__(
        self,
        stream: AsyncGenerator[AssistantStreamChunk, None],
        heartbeat: HeartbeatOption = True,
    ):
        super().__init__(stream, AssistantTransportEncoder(), heartbeat=heartbeat)

File: packages/assistant-stream/src/core/serialization/assistant-transport/AssistantTransport.ts
import sjson from "secure-json-parse";
import type { AssistantStreamChunk } from "../../AssistantStreamChunk";
import { PipeableTransformStream } from "../../utils/stream/PipeableTransformStream";
import {
  createSSEJsonDecoder,
  createSSEJsonEncoder,
  SSE_HEADERS,
} from "../../utils/stream/SSEJson";
import type { AssistantStreamEncoder } from "../../AssistantStream";

type ChunkFields = Record<string, unknown>;

type ChunkRule = {
  kind: "message" | "part-addressed";
  valid: (chunk: ChunkFields) => boolean;
};

const noFields = () => true;
const requiredObject = (key: string) => (c: ChunkFields) => {
  const value = c[key];
  return typeof value === "object" && value !== null && !Array.isArray(value);
};
const requiredString = (key: string) => (c: ChunkFields) =>
  typeof c[key] === "string";
const requiredArray = (key: string) => (c: ChunkFields) =>
  Array.isArray(c[key]);
const optionalBoolean = (key: string) => (c: ChunkFields) =>
  c[key] === undefined || typeof c[key] === "boolean";

const KNOWN_CHUNK_TYPES: Record<AssistantStreamChunk["type"], ChunkRule> = {
  "part-start": { kind: "message", valid: requiredObject("part") },
  "part-finish": { kind: "part-addressed", valid: noFields },
  "tool-call-args-text-finish": { kind: "part-addressed", valid: noFields },
  "text-delta": { kind: "part-addressed", valid: requiredString("textDelta") },
  annotations: { kind: "message", valid: requiredArray("annotations") },
  data: { kind: "message", valid: requiredArray("data") },
  "step-start": { kind: "message", valid: noFields },
  "step-finish": { kind: "message", valid: requiredString("finishReason") },
  "message-finish": { kind: "message", valid: requiredString("finishReason") },
  result: { kind: "part-addressed", valid: optionalBoolean("isError") },
  error: { kind: "message", valid: noFields },
  "update-state": { kind: "message", valid: requiredArray("operations") },
};

const parseChunk = (data: string): AssistantStreamChunk | string => {
  let value: unknown;
  try {
    value = sjson.parse(data);
  } catch {
    return "unparseable";
  }
  if (typeof value !== "object" || value === null || Array.isArray(value))
    return "not-an-object";
  const { type, path } = value as { type?: unknown; path?: unknown };
  if (
    typeof type !== "string" ||
    !Object.prototype.hasOwnProperty.call(KNOWN_CHUNK_TYPES, type)
  )
    return "unknown-type";
  const rule = KNOWN_CHUNK_TYPES[type as AssistantStreamChunk["type"]];
  if (!rule.valid(value as Record<string, unknown>))
    return `invalid-fields:${type}`;
  if (path === undefined) {
    if (rule.kind !== "message") return `missing-path:${type}`;
    return { ...value, path: [] } as unknown as AssistantStreamChunk;
  }
  if (
    !Array.isArray(path) ||
    !path.every((entry) => Number.isInteger(entry) && entry >= 0)
  )
    return "invalid-path";
  return value as AssistantStreamChunk;
};

/**
 * AssistantTransportEncoder encodes AssistantStreamChunks into SSE format
 * and emits [DONE] when the stream completes.
 */
export class AssistantTransportEncoder
  extends PipeableTransformStream<AssistantStreamChunk, Uint8Array<ArrayBuffer>>
  implements AssistantStreamEncoder
{
  headers = new Headers(SSE_HEADERS);

  constructor() {
    super(createSSEJsonEncoder<AssistantStreamChunk>("[DONE]"));
  }
}

/**
 * AssistantTransportDecoder decodes SSE format into AssistantStreamChunks.
 * It stops decoding when it encounters [DONE].
 */
export class AssistantTransportDecoder extends PipeableTransformStream<
  Uint8Array<ArrayBuffer>,
  AssistantStreamChunk
> {
  constructor(options: { strict?: boolean | undefined } = {}) {
    const strict = options.strict ?? true;
    const warnedReasons = new Set<string>();
    super(
      createSSEJsonDecoder<AssistantStreamChunk>({
        parse(data, controller) {
          const chunk = parseChunk(data);
          if (typeof chunk === "string") {
            if (!warnedReasons.has(chunk)) {
              warnedReasons.add(chunk);
              console.warn(
                `Dropped invalid assistant-transport chunk (${chunk}): ${data.slice(0, 200)}`,
              );
            }
          } else {
            controller.enqueue(chunk);
          }
        },
        done: {
          marker: "[DONE]",
          onMissing() {
            if (strict) {
              throw new Error(
                "Stream ended abruptly without receiving [DONE] marker",
              );
            }
            console.warn(
              "Stream ended abruptly without receiving [DONE] marker",
            );
          },
        },
        ...(strict
          ? { strict: true }
          : {
              strict: false,
              onUnknownEvent(event) {
                if (!warnedReasons.has(`event:${event}`)) {
                  warnedReasons.add(`event:${event}`);
                  console.error(`Ignored unknown SSE event type: ${event}`);
                }
              },
            }),
      }),
    );
  }
}
